core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/plangen/LogicalPlan.java

```diff
@@ -103,24 +111,90 @@ public LogicalPlan(
      * @param operatorID, the ID of an operator
      * @return Schema, which includes the attributes setting of the operator
      */
-    public Schema getOperatorOutputSchema(String operatorID) throws PlanGenException {
+    public Schema getOperatorOutputSchema(String operatorID) throws PlanGenException, DataflowException {
 
         if (UPDATED) {
             buildOperators();
             checkGraphCyclicity();
-            checkSourceOperator();
 
             connectOperators(operatorObjectMap);
         }
-        IOperator currentOperator = operatorObjectMap.get(operatorID);
 
+        IOperator currentOperator = operatorObjectMap.get(operatorID);
         currentOperator.open();
         Schema operatorSchema = currentOperator.getOutputSchema();
         currentOperator.close();
 
         return operatorSchema;
     }
 
+    /**
+     * Updates the current plan and fetch the schema from an operator
+     * @param operatorID, the ID of an operator
+     * @return Schema, which includes the attributes setting of the operator
+     */
+    public Schema getOperatorOutputSchema(String operatorID, Map<String, List<Schema>> inputSchemas)
+            throws PlanGenException, DataflowException {
+
+        IOperator currentOperator = operatorObjectMap.get(operatorID);
+        Schema outputSchema = null;
+        if (currentOperator instanceof ISourceOperator) {
+            outputSchema = currentOperator.transformToOutputSchema();
+        } else if (inputSchemas.containsKey(operatorID)) {
+            List<Schema> inputSchema = inputSchemas.get(operatorID);
+            try {
+                outputSchema = currentOperator.transformToOutputSchema(
+                        inputSchema.toArray(new Schema[inputSchema.size()]));
+            } catch (TexeraException e) {
+                System.out.println(e.getMessage());
+            }
+        }
+        return outputSchema;
+    }
+
+    public Map<String, List<Schema>> retrieveAllOperatorInputSchema() throws PlanGenException {
+
+        if (UPDATED) {
```

Comment:
AFAIK every request from client will make a new LogicalPlan inside QueryPlanResource.java and therefore UPDATED will always be true here. Once buildOperators() is executed, UPDATED will always be false.